File: qoogear/qoogear-sdk/cpp/src/security/chip_auth.cpp
#include "qoogear/security/cert_verifier.h"
#include "qoogear/security/chip_auth.h"
#include <chrono>
#include <random>
#include <sstream>
#include <iomanip>

namespace qoogear {
namespace security {

// ---- CertVerifier ----

CertVerifier::CertVerifier(bool cache_enabled, uint32_t cache_ttl_seconds)
    : cache_enabled_(cache_enabled), cache_ttl_(cache_ttl_seconds) {}

CertInfo CertVerifier::verify_cert_hash(const std::string& cert_hash) {
    std::string cache_key = "hash:" + cert_hash;
    if (cache_enabled_) {
        auto it = cache_.find(cache_key);
        if (it != cache_.end()) return it->second;
    }

    bool is_valid = cert_hash.length() >= 16 && cert_hash.substr(0, 3) == "MFQ";
    CertInfo info;
    info.cert_number = is_valid ? ("MFQ-2026-" + cert_hash.substr(0, 8)) : "";
    info.cert_level = "BASIC";
    info.product_name = "Unknown Product";
    info.vendor_name = "Unknown Vendor";
    info.issued_at = "2026-01-01T00:00:00Z";
    info.expires_at = "2027-01-01T00:00:00Z";
    info.is_valid = is_valid;
    info.is_expired = false;
    info.is_revoked = false;

    if (cache_enabled_) cache_[cache_key] = info;
    return info;
}

CertInfo CertVerifier::verify_cert_number(const std::string& cert_number) {
    // 桩实现
    CertInfo info;
    info.cert_number = cert_number;
    info.is_valid = cert_number.find("MFQ") == 0;
    return info;
}

bool CertVerifier::verify_signature(const std::vector<uint8_t>& payload,
                                    const std::vector<uint8_t>& signature,
                                    const std::string& public_key_pem) {
    if (payload.empty() || signature.empty()) return false;
    return signature.size() >= 32;
}

std::string CertVerifier::compute_cert_hash(const std::string& cert_json) {
    // 简化桩实现
    std::hash<std::string> hasher;
    size_t hash = hasher(cert_json);
    std::ostringstream oss;
    oss << std::hex << hash;
    return oss.str();
}

void CertVerifier::clear_cache() {
    cache_.clear();
}

// ---- ChipAuthenticator ----

ChipAuthenticator::ChipAuthenticator(uint8_t i2c_address)
    : i2c_address_(i2c_address) {
    // 生成随机密钥
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<uint8_t> dist(0, 255);
    secret_key_.resize(32);
    for (auto& b : secret_key_) b = dist(gen);
}

bool ChipAuthenticator::probe() {
    connected_ = true;
    auto now = std::chrono::system_clock::now().time_since_epoch().count();

    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<uint32_t> dist(0, 0xFFFFFFFF);

    std::ostringstream chip_oss;
    chip_oss << "CHIP-" << std::hex << dist(gen);
    chip_info_.chip_id = chip_oss.str();

    std::ostringstream sn_oss;
    sn_oss << "SN-" << std::dec << (now / 1000000000);
    chip_info_.chip_serial = sn_oss.str();
    chip_info_.certificate_id = "MFQ-2026-BASIC-00001";
    chip_info_.batch_number = "BATCH-2026-001";
    chip_info_.status = ChipStatus::ACTIVE;
    chip_info_.burned_at = static_cast<uint64_t>(now);
    return true;
}

bool ChipAuthenticator::challenge_response() {
    if (!connected_) return false;
    return true;
}

bool ChipAuthenticator::mutual_authenticate() {
    return challenge_response();
}

ChipInfo ChipAuthenticator::read_chip_info() {
    return chip_info_;
}

std::string ChipAuthenticator::read_certificate() {
    if (!connected_) return "";
    return "{\"cert_number\":\"" + chip_info_.certificate_id + "\",\"status\":\"active\"}";
}

std::vector<uint8_t> ChipAuthenticator::read_public_key() {
    if (!connected_) return {};
    return secret_key_;
}

bool ChipAuthenticator::provision(const std::string& certificate_id,
                                  const std::vector<uint8_t>& secret_key) {
    if (!connected_) return false;
    if (chip_info_.status != ChipStatus::BLANK) return false;
    if (!secret_key.empty()) secret_key_ = secret_key;
    chip_info_.certificate_id = certificate_id;
    chip_info_.status = ChipStatus::PROVISIONED;
    return true;
}

bool ChipAuthenticator::activate() {
    if (!connected_) return false;
    if (chip_info_.status != ChipStatus::PROVISIONED) return false;
    chip_info_.status = ChipStatus::ACTIVE;
    return true;
}

bool ChipAuthenticator::revoke() {
    if (!connected_) return false;
    chip_info_.status = ChipStatus::REVOKED;
    return true;
}

uint32_t ChipAuthenticator::get_usage_counter() {
    auto now = std::chrono::system_clock::now().time_since_epoch().count();
    return static_cast<uint32_t>(now % 100000);
}

bool ChipAuthenticator::increment_counter() {
    if (!connected_) return false;
    return true;
}

} // namespace security
} // namespace qoogear

File: qoobot-service/gear/qoogear-sdk/python/qoogear_sdk/utils/chip_auth.py
"""MFQ 认证芯片通信"""

from __future__ import annotations

import hashlib
import hmac
import os
import time
from dataclasses import dataclass
from enum import Enum
from typing import Optional


class ChipStatus(Enum):
    BLANK = "blank"
    PROVISIONED = "provisioned"
    ACTIVE = "active"
    REVOKED = "revoked"
    EXPIRED = "expired"


@dataclass
class ChipInfo:
    """认证芯片信息"""
    chip_id: str = ""
    chip_serial: str = ""
    certificate_id: str = ""
    batch_number: str = ""
    status: ChipStatus = ChipStatus.BLANK
    burned_at: float = 0.0


class ChipAuthenticator:
    """MFQ 认证芯片通信器。

    通过 I2C/1-Wire 与配件上的 MFQ 认证芯片通信。
    实现挑战-响应认证、证书读取和芯片生命周期管理。

    Usage:
        auth = ChipAuthenticator()
        if auth.probe():
            chip_info = auth.read_chip_info()
            is_authentic = auth.challenge_response()
    """

    # 模拟的 MFQ 认证芯片 I2C 地址
    DEFAULT_I2C_ADDRESS = 0x50

    def __init__(self, i2c_address: int = DEFAULT_I2C_ADDRESS):
        self._i2c_address = i2c_address
        self._connected = False
        self._chip_info = ChipInfo()
        self._secret_key = os.urandom(32)

    @property
    def is_connected(self) -> bool:
        return self._connected

    @property
    def chip_info(self) -> ChipInfo:
        return self._chip_info

    # ---- 芯片探测 ----

    def probe(self) -> bool:
        """探测认证芯片是否存在"""
        # 桩实现：模拟芯片存在
        self._connected = True
        self._chip_info = ChipInfo(
            chip_id="CHIP-" + os.urandom(4).hex().upper(),
            chip_serial=f"SN-{int(time.time())}",
            certificate_id="MFQ-2026-BASIC-00001",
            batch_number="BATCH-2026-001",
            status=ChipStatus.ACTIVE,
            burned_at=time.time(),
        )
        return True

    # ---- 挑战-响应认证 ----

    def challenge_response(self) -> bool:
        """执行挑战-响应认证协议

        1. 主机生成随机挑战码
        2. 芯片用内置密钥对挑战码签名
        3. 主机验证签名
        """
        if not self._connected:
            return False

        # 生成挑战码
        challenge = os.urandom(32)

        # 芯片端：用密钥签名（模拟）
        chip_signature = hmac.new(self._secret_key, challenge, hashlib.sha256).digest()

        # 主机端：验证签名
        expected = hmac.new(self._secret_key, challenge, hashlib.sha256).digest()
        return hmac.compare_digest(chip_signature, expected)

    def mutual_authenticate(self) -> bool:
        """双向认证：同时验证主机和芯片"""
        if not self._connected:
            return False
        return self.challenge_response()

    # ---- 芯片数据读取 ----

    def read_chip_info(self) -> Optional[ChipInfo]:
        """读取芯片信息"""
        if not self._connected:
            return None
        return self._chip_info

    def read_certificate(self) -> Optional[str]:
        """读取存储在芯片中的证书"""
        if not self._connected:
            return None
        return f'{{"cert_number":"{self._chip_info.certificate_id}","status":"active"}}'

    def read_public_key(self) -> Optional[bytes]:
        """读取芯片公钥"""
        if not self._connected:
            return None
        return hashlib.sha256(self._secret_key).digest()

    # ---- 芯片生命周期 ----

    def provision(self, certificate_id: str, secret_key: Optional[bytes] = None) -> bool:
        """烧录证书到芯片"""
        if not self._connected:
            return False
        if self._chip_info.status != ChipStatus.BLANK:
            return False

        if secret_key:
            self._secret_key = secret_key

        self._chip_info.certificate_id = certificate_id
        self._chip_info.status = ChipStatus.PROVISIONED
        self._chip_info.burned_at = time.time()
        return True

    def activate(self) -> bool:
        """激活芯片"""
        if not self._connected:
            return False
        if self._chip_info.status != ChipStatus.PROVISIONED:
            return False
        self._chip_info.status = ChipStatus.ACTIVE
        return True

    def revoke(self) -> bool:
        """吊销芯片"""
        if not self._connected:
            return False
        self._chip_info.status = ChipStatus.REVOKED
        return True

    # ---- 安全计数器 ----

    def get_usage_counter(self) -> int:
        """读取使用计数器（防克隆）"""
        if not self._connected:
            return 0
        return int(time.time()) % 100000

    def increment_counter(self) -> bool:
        """递增使用计数器"""
        if not self._connected:
            return False
        return True
